File: leetcode_problems/python/lc98_validate_bst.py
from utils import TreeNode
from utils import build_tree
from utils import print_tree


class Solution:
    def isValidBST(self, root: TreeNode | None) -> bool:
        def validate(node: TreeNode | None, low: float, high: float) -> bool:
            if not node:
                return True
            if not (low < node.val < high):
                return False
            return validate(node.left, low, node.val) and validate(
                node.right,
                node.val,
                high,
            )

        return validate(root, float("-inf"), float("inf"))


if __name__ == "__main__":
    arr = [5, 1, 4, None, None, 3, 6]
    arr = [2, 1, 3]
    arr = [1, None, 1]
    arr = [5, 4, 6, None, None, 3, 7]
    root = build_tree(arr)
    print_tree(root)
    print(Solution().isValidBST(root))

File: leetcode_problems/go/lc98_validate_bst.go
package main

import "math"

func isValidBSTHelper(node *TreeNode, mini int, maxi int) bool {
	if node == nil {
		return true
	}
	if node.Val <= mini || node.Val >= maxi {
		return false
	}
	return isValidBSTHelper(node.Left, mini, node.Val) && isValidBSTHelper(node.Right, node.Val, maxi)
}

func isValidBST(root *TreeNode) bool {
	return isValidBSTHelper(root, math.MinInt64, math.MaxInt64)
}
